README.md
![](https://d362q4tvy1elxj.cloudfront.net/header_logoheader.png)

PHP-SDK
==========

Table of Contents
=================

  * [Table of contents](#table-of-contents)
  * [Installation](#installation)
  * [Getting Started](#getting-started)
    * [Retrieve a Specific Template](#retrieve-a-specific-template)
    * [List all Signed Waivers](#list-all-signed-waivers)
    * [Retrieve a Specific Waiver](#retrieve-a-specific-waiver)
    * [Retrieve Photos on a Waiver](#retrieve-photos-on-a-waiver)
    * [Search for Waivers](#search-for-waivers)
    * [Retrieve/Set Webhook Config](#retrieveset-webhook-configuration)
    * [Retrieve Messages From Webhook Queues](#retrieve-messages-from-webhook-queues)
  * [Exception Handling](#exception-handling)
    * [Status Codes](#status-codes)
  * [Advanced](#advanced)
    * [Raw Responses](#raw-responses)
    * [URL Generation](#url-generation)
    * [Authentication](#authentication)
 * [API Documentation](#api-documentation)
    * [Smartwaiver/Exceptions/SmartwaiverHTTPException](#smartwaiverexceptionssmartwaiverhttpexception)
    * [Smartwaiver/Exceptions/SmartwaiverRateLimitException](#smartwaiverexceptionssmartwaiverratelimitexception)
    * [Smartwaiver/Exceptions/SmartwaiverSDKException](#smartwaiverexceptionssmartwaiversdkexception)
    * [Smartwaiver/Smartwaiver](#smartwaiversmartwaiver)
    * [Smartwaiver/SmartwaiverRawResponse](#smartwaiversmartwaiverrawresponse)
    * [Smartwaiver/SmartwaiverResponse](#smartwaiversmartwaiverresponse)
    * [Smartwaiver/SmartwaiverRoutes](#smartwaiversmartwaiverroutes)
    * [Smartwaiver/Types/Data/SmartwaiverTemplateData](#smartwaivertypesdatasmartwaivertemplatedata)
    * [Smartwaiver/Types/SmartwaiverCustomField](#smartwaivertypessmartwaivercustomfield)
    * [Smartwaiver/Types/SmartwaiverDynamicProcess](#smartwaivertypessmartwaiverdynamicprocess)
    * [Smartwaiver/Types/SmartwaiverDynamicTemplate](#smartwaivertypessmartwaiverdynamictemplate)
    * [Smartwaiver/Types/SmartwaiverFlag](#smartwaivertypessmartwaiverflag)
    * [Smartwaiver/Types/SmartwaiverGuardian](#smartwaivertypessmartwaiverguardian)
    * [Smartwaiver/Types/SmartwaiverParticipant](#smartwaivertypessmartwaiverparticipant)
    * [Smartwaiver/Types/SmartwaiverPhoto](#smartwaivertypessmartwaiverphoto)
    * [Smartwaiver/Types/SmartwaiverPhotos](#smartwaivertypessmartwaiverphotos)
    * [Smartwaiver/Types/SmartwaiverSearch](#smartwaivertypessmartwaiversearch)
    * [Smartwaiver/Types/SmartwaiverSignatures](#smartwaivertypessmartwaiversignatures)
    * [Smartwaiver/Types/SmartwaiverTemplate](#smartwaivertypessmartwaivertemplate)
    * [Smartwaiver/Types/SmartwaiverType](#smartwaivertypessmartwaivertype)
    * [Smartwaiver/Types/SmartwaiverWaiver](#smartwaivertypessmartwaiverwaiver)
    * [Smartwaiver/Types/SmartwaiverWaiverSummary](#smartwaivertypessmartwaiverwaiversummary)
    * [Smartwaiver/Types/SmartwaiverWebhook](#smartwaivertypessmartwaiverwebhook)
    * [Smartwaiver/Types/Template/SmartwaiverTemplateBody](#smartwaivertypestemplatesmartwaivertemplatebody)
    * [Smartwaiver/Types/Template/SmartwaiverTemplateCompletion](#smartwaivertypestemplatesmartwaivertemplatecompletion)
    * [Smartwaiver/Types/Template/SmartwaiverTemplateConfig](#smartwaivertypestemplatesmartwaivertemplateconfig)
    * [Smartwaiver/Types/Template/SmartwaiverTemplateElectronicConsent](#smartwaivertypestemplatesmartwaivertemplateelectronicconsent)
    * [Smartwaiver/Types/Template/SmartwaiverTemplateGuardian](#smartwaivertypestemplatesmartwaivertemplateguardian)
    * [Smartwaiver/Types/Template/SmartwaiverTemplateHeader](#smartwaivertypestemplatesmartwaivertemplateheader)
    * [Smartwaiver/Types/Template/SmartwaiverTemplateMeta](#smartwaivertypestemplatesmartwaivertemplatemeta)
    * [Smartwaiver/Types/Template/SmartwaiverTemplateParticipants](#smartwaivertypestemplatesmartwaivertemplateparticipants)
    * [Smartwaiver/Types/Template/SmartwaiverTemplateProcessing](#smartwaivertypestemplatesmartwaivertemplateprocessing)
    * [Smartwaiver/Types/Template/SmartwaiverTemplateSignatures](#smartwaivertypestemplatesmartwaivertemplatesignatures)
    * [Smartwaiver/Types/Template/SmartwaiverTemplateStandardQuestions](#smartwaivertypestemplatesmartwaivertemplatestandardquestions)
    * [Smartwaiver/Types/Template/SmartwaiverTemplateStyling](#smartwaivertypestemplatesmartwaivertemplatestyling)
    * [Smartwaiver/Types/WebhookQueues/SmartwaiverWebhookMessage](#smartwaivertypeswebhookqueuessmartwaiverwebhookmessage)
    * [Smartwaiver/Types/WebhookQueues/SmartwaiverWebhookMessageDelete](#smartwaivertypeswebhookqueuessmartwaiverwebhookmessagedelete)
    * [Smartwaiver/Types/WebhookQueues/SmartwaiverWebhookMessagePayload](#smartwaivertypeswebhookqueuessmartwaiverwebhookmessagepayload)
    * [Smartwaiver/Types/WebhookQueues/SmartwaiverWebhookQueue](#smartwaivertypeswebhookqueuessmartwaiverwebhookqueue)
    * [Smartwaiver/Types/WebhookQueues/SmartwaiverWebhookQueues](#smartwaivertypeswebhookqueuessmartwaiverwebhookqueues)

Installation
==========

To add the Smartwaiver SDK to your [Composer](https://getcomposer.org/) project:

    composer require smartwaiver/smartwaiver-sdk

Note: Smartwaiver SDK depends upon Guzzle so if you are not using composer you will need to handle those dependencies yourself.

Getting Started
==========

All that is required to start using the SDK is a Smartwaiver account and the API Key for that account.
In all of the examples you will need to put the API Key into the code wherever it says: `[INSERT API KEY]`

It's time to start making requests.
A good first request is to list all waiver templates for your account.
Here is the code to do that:

```php
// The API Key for your account
$apiKey = '[INSERT API KEY]';

// Set up your Smartwaiver connection using your API Key
$sw = new Smartwaiver($apiKey);

// Now request a list of all the waiver templates
$templates = $sw->getWaiverTemplates();
```

That's it! You've just requested all waiver templates in your account.
But, now it's time to do something with them.
Let's loop through those templates and print out the ID and Title of each template:

```php
foreach ($templates as $template) {
    echo $template->templateId . ': ' . $template->title . PHP_EOL;
}
```

Awesome! For more details on all the different properties a waiver template has, check out [TemplateProperties.php](examples/templates/TemplateProperties.php)

Now that you've got your first request, check out the sections below to accomplish specific actions.

Retrieve a Specific Template
---------

First let's set up the basic Smartwaiver object.
Make sure to put in your account's API Key where it says `[INSERT API KEY]`

```php
// The API Key for your account
$apiKey = '[INSERT API KEY]';

// Set up your Smartwaiver connection using your API Key
$sw = new Smartwaiver($apiKey);
```

Now we can request information about a specific template.
To do this we need the template ID.
If you don't know a template ID for your account, try listing all waiver templates for you account, as shown [here](#getting-started), and copying one of the ID's that is printed out.
Once we have a template ID we can execute a request to get the information about the template:

```php
// The unique ID of the template to be retrieved
$templateId = '[INSERT TEMPLATE ID]';

// Retrieve a specific template (SmartwaiverTemplate object)
$template = $sw->getWaiverTemplate($templateId);
```

Now let's print out some information about this template.

```php
// Access properties of the template
echo PHP_EOL . 'List single template:' . PHP_EOL;
echo $template->templateId . ': ' . $template->title . PHP_EOL;
```

To see all the different properties a waiver template has, check out [TemplateProperties.php](examples/templates/TemplateProperties.php)

List All Signed Waivers
----------

First let's set up the basic Smartwaiver object. Make sure to put in your account's API Key where it says `[INSERT API KEY]`

```php
// The API Key for your account
$apiKey = '[INSERT API KEY]';

// Set up your Smartwaiver connection using your API Key
$sw = new Smartwaiver($apiKey);
```

Now we can request signed waivers from your account.

```php
// Get a list of summaries of waivers
$waiverSummaries = $sw->getWaiverSummaries();
```

With this done, we can iterate over the returned summaries to see what is stored.
The default limit is 20, which means if you have more than 20 in your account, only the most recent 20 will be returned

```php
// Loop through the waivers and access their properties
echo 'List all waivers:' . PHP_EOL;
foreach ($waiverSummaries as $waiverSummary) {
    echo $waiverSummary->waiverId . ': ' . $waiverSummary->title . PHP_EOL;
}
```

To see all the different properties a waiver summary has, check out [WaiverSummaryProperties.php](examples/waivers/WaiverSummaryProperties.php)

Once we have a waiver summary, we can access all the detailed information about the waiver. To do that look [here](#retrieve-a-specific-waiver).

But, we can also restrict our query with some parameters.
For example, what if we only want to return 5 waivers, (the default is 20).
Here is the code to do that:

```php
// Set the limit
$limit = 5;

// Get a list of summaries of waivers
$waiverSummaries = $sw->getWaiverSummaries($limit);
```

Or what if we only want any waivers that have not been verified (either by email or at the kiosk)?

```php
// Set the limit
$limit = 5;

// Set the verified parameter
$verified = false;

// Get a list of summaries of waivers
$waiverSummaries = $sw->getWaiverSummaries($limit, $verified);
```

What other parameters can you use? Here is an example using all of them:

```php
// An example limiting the parameters
$limit = 5;                                     // Limit number returned to 5
$verified = true;                               // Limit only to waivers that were signed at a kiosk or verified over email
$templateId = '[INSERT TEMPLATE ID]';           // Limit query to waivers of this template ID
$fromDts = date('c', strtotime('2016-11-01'));  // Limit to waivers signed in November of 2016
$toDts = date('c', strtotime('2016-12-01'));
$firstName = 'Kyle';                            // Limit to waivers with a participant named Kyle Smith
$lastName = 'Smith';

// Get a list of summaries of waivers
$waiverSummaries = $sw->getWaiverSummaries($limit, $verified, $templateId, $fromDts, $toDts);
```

These examples are also available in [ListAllWaivers.php](examples/waivers/ListAllWaivers.php)

### Parameter Options

| Parameter Name | Default Value | Accepted Values   | Notes                                                                                   |
| -------------- | ------------- | ----------------- | --------------------------------------------------------------------------------------- |
| limit          | 20            | 1 - 100           | Limit number of returned waivers                                                        |
| verified       | null          | true/false/null   | Limit selection to waiver that have been verified (true), not (false), or both (null)   |
| templateId     |               | Valid Template ID | Limit signed waivers to only this template                                              |
| fromDts        |               | ISO 8601 Date     | Limit to signed waivers between from and to dates (requires toDts)                      |
| toDts          |               | ISO 8601 Date     | Limit to signed waivers between from and to dates (requires fromDts)                    |
| firstName      |               | Alpha-numeric     | Limit to signed waivers that have a participant with this first name (Case Insensitive) |
| lastName       |               | Alpha-numeric     | Limit to signed waivers that have a participant with this last name (Case Insensitive)  |

Retrieve a Specific Waiver
----------

What if we want to retrieve a specific waiver?
All we need for that is a waiver ID.
If you don't have a waiver ID to use, you can get a list of signed waivers in your account [here](#list-all-signed-waivers)

First let's set up the basic Smartwaiver object. Make sure to put in your account's API Key where it says `[INSERT API KEY]`

```php
// The API Key for your account
$apiKey = '[INSERT API KEY]';

// Set up your Smartwaiver connection using your API Key
$sw = new Smartwaiver($apiKey);
```

Now, we can request the information about a specific waiver.
Make sure to put your waiver ID in where it says `[INSERT WAIVER ID]`

```php
// The unique ID of the signed waiver to be retrieved
$waiverId = '[INSERT WAIVER ID]';

// Get a specific waiver
$waiver = $sw->getWaiver($waiverId);
```

The waiver object has many different properties that can be accessed.
For example, we can print out the waiver ID and title of the waiver.

```php
// Access properties of waiver
echo PHP_EOL . 'List single waiver:' . PHP_EOL;
echo $waiver->waiverId . ': ' . $waiver->title . PHP_EOL;
```

To see a full list of all properties that a waiver object contains, check out [WaiverProperties.php](examples/waivers/WaiverProperties.php)

We can also request that the PDF of the signed waiver as a Base 64 Encoded string be included. Here is the request to do that:

```php
// The unique ID of the signed waiver to be retrieved
$waiverId = '[INSERT WAIVER ID]';

$pdf = true;

// Get the waiver object
$waiver = $sw->getWaiver($waiverId, $pdf);
```

The code provided here is also combined in to one example in [RetrieveSingleWaiver.php](examples/waivers/RetrieveSingleWaiver.php)

Retrieve Photos on a Waiver
----------

We can also use the API to retrieve any photos taken when the waiver was signed or attached later with the console.
All we need is you're API key and the ID of a signed waiver, which has attached photos.


If you don't have a waiver ID to use, you can get a list of signed waivers in your account [here](#list-all-signed-waivers)

First let's set up the basic Smartwaiver object. Make sure to put in your account's API Key where it says `[INSERT API KEY]`

```php
// The API Key for your account
$apiKey = '[INSERT API KEY]';

// Set up your Smartwaiver connection using your API Key
$sw = new Smartwaiver($apiKey);
```

Now, we can request the photos on a specific waiver.
Make sure to put your waiver ID in where it says `[INSERT WAIVER ID]`

```php
// TThe unique ID of the signed waiver to retrieve the photos for
$waiverId = '[INSERT WAIVER ID]';

// Get the photos for a specific waiver
$photos = $sw->getWaiverPhotos($waiverId);
```

This photos object has a little meta-data we can print out:

```php
// Print a little header
echo PHP_EOL . 'Waiver Photos for: ' . $photos->title . PHP_EOL;
// echo $photos->waiverId;
// echo $photos->templateId;
// echo $photos->createdOn;
```

Using this photos object, we can loop through all the photos and process them however we like:

```php
// Loop through photos and print out some meta-data
foreach ($photos->photos as $photo) {
    echo $photo->photoId . ': ' . $photo->date;
    // Other fields
    // echo $photo->type;
    // echo $photo->tag;
    // echo $photo->fileType;
    // echo $photo->photo; // Base 64 encoded photo
}
```

The code provided here is also combined in to one example in [RetrieveWaiverPhotos.php](examples/waivers/RetrieveWaiverPhotos.php)

Search for Waivers
----------

First let's set up the basic Smartwaiver object. Make sure to put in your account's API Key where it says `[INSERT API KEY]`

```php
// The API Key for your account
$apiKey = '[INSERT API KEY]';

// Set up your Smartwaiver connection using your API Key
$sw = new Smartwaiver($apiKey);
```

Now we can request a search for signed waivers from your account.

```php
// Request all waivers signed in 2017
$search = $sw->search('', '2017-01-01 00:00:00');
```

**Note: The search route is a blocking search. Thus, a request to search for large amounts of data can take up to a few seconds.
As such, this route should not be used for anything where real-time performance is important. Instead use the Waivers route.**

This will return a search object containing metadata about the results of our search.
We can easily print out all that information:

```php
// Print out some information about the result of the search
echo 'Search Complete:' . PHP_EOL;
echo "\t" . 'Search ID: ' . $search->guid . PHP_EOL;
echo "\t" . 'Waiver Count: ' . $search->count . PHP_EOL;
echo "\t" . $search->pages . ' pages of size ' . $search->pageSize . PHP_EOL . PHP_EOL;
```

The server has stored the results of our search request under the GUID given.
We can now loop through the pages and request each page, which will be a list of up to 100 waivers.
For example, if we wanted to created a list of all first names from our search, we would do that like this:

```php
// First we set up our list
$nameList = [];

// Loop through all the pages in the search result
for($i = 0; $i < $search->pages; $i++) {
    echo 'Requesting page: ' . $i . '/' . $search->pages . '...' . PHP_EOL;

    // Request each page from the server
    $waivers = $sw->searchResult($search, $i);

    echo 'Processing page: ' . $i . '/' . $search->pages . '...' . PHP_EOL;

    // Loop through the waivers and create a comma separated list of first names
    foreach ($waivers as $waiver) {
        array_push($nameList, $waiver->firstName);

        // View all accessible properties of a waiver object in:
        // examples/waivers/WaiverProperties.php
    }
}
```

To see all the different properties a waiver has, check out [WaiverProperties.php](examples/waivers/WaiverProperties.php)

This examples is also available in [BasicSearch.php](examples/search/BasicSearch.php)

### Search Parameters

We can also restrict our search with more parameters.
For example, what if we only want to return waivers for one of the templates in our account.
Here is the code to do that:

```php
// The unique ID of the template to search for
$templateId = '[INSERT TEMPLATE ID]';

// Request all waivers signed for this template
$search = $sw->search($templateId);
```

Or what if we only want any waivers that have not been verified (either by email or at the kiosk)?

```php
// Request all waivers signed that not have been email verified
$search = $sw->search('', '', '', '', '', false);
```

What other parameters can you use? Here are some more examples:

```php
// Request all waivers signed for this template after the given date
$search = $sw->search($templateId, '2017-01-01 00:00:00');

// Request all waivers signed for this template before the given date
$search = $sw->search($templateId, '', '2017-01-01 00:00:00');

// Request all waivers signed for this template with a participant name Kyle
$search = $sw->search($templateId, '', '', 'Kyle');

// Request all waivers signed for this template with a participant name Kyle Smith
$search = $sw->search($templateId, '', '', 'Kyle', 'Smith');

// Request all waivers signed with a participant name Kyle that have been email verified
$search = $sw->search('', '', '', 'Kyle', '', true);

// Request all waivers signed in ascending sorted order
//$search = $sw->search($templateId, '', '', '', '', null, false);
```

These examples are also available in [SearchParams.php](examples/search/SearchParams.php)

### Parameter Options

| Parameter Name | Default Value | Accepted Values   | Notes                                                                                   |
| -------------- | ------------- | ----------------- | --------------------------------------------------------------------------------------- |
| templateId     |               | Valid Template ID | Limit signed waivers to only this template                                              |
| fromDts        |               | ISO 8601 Date     | Limit to signed waivers between after this date                                         |
| toDts          |               | ISO 8601 Date     | Limit to signed waivers between before this date                                        |
| firstName      |               | Alpha-numeric     | Limit to signed waivers that have a participant with this first name (Case Insensitive) |
| lastName       |               | Alpha-numeric     | Limit to signed waivers that have a participant with this last name (Case Insensitive)  |
| verified       | null          | true/false/null   | Limit selection to waiver that have been verified (true), not (false), or both (null)   |
| sortDesc       | true          | true/false        | Sort results in descending (latest signed waiver first) order                           |


Retrieve/Set Webhook Configuration
----------

You can both retrieve and set your account's webhook configuration through this SDK with a couple simple calls.
To view your current webhook settings, we first need to set a Smartwaiver object.
Make sure to put in your account's API Key where it says `[INSERT API KEY]`

```php
// The API Key for your account
$apiKey = '[INSERT API KEY]';

// Set up your Smartwaiver connection using your API Key
$sw = new Smartwaiver($apiKey);
```

Now, it's easy to request the webhook configuration:

```php
// Get the current webhook settings
$webhooks = $sw->getWebhookConfig();
```

And, now we can print out the information:

```php
// Access the webhook config
echo 'Endpoint: ' . $webhooks->endpoint . PHP_EOL;
echo 'EmailValidationRequired: ' . $webhooks->emailValidationRequired . PHP_EOL;
```

The Email Validation Required is whether the webhook will fire before, after, or before and after a waiver is verified.
The endpoint is simply the endpoint URL for the webhook.

And changing your webhook configuration is just as easy.
The new configuration will be returned from the request and can be access just like the read request above.

```php
// The new values to set
$endpoint = 'http://endpoint.example.org';
$emailValidationRequired = SmartwaiverWebhook::WEBHOOK_AFTER_EMAIL_ONLY;

// Set the webhook to new values
$webhook = $sw->setWebhookConfig($endpoint, $emailValidationRequired);

// Access the new webhook config
echo 'Successfully set new configuration.' . PHP_EOL;
echo 'Endpoint: ' . $newWebhook->endpoint . PHP_EOL;
echo 'EmailValidationRequired: ' . $newWebhook->emailValidationRequired . PHP_EOL;
```

If you wish to remove a webhook completely from the account you may also do that like this:

```php
$sw->deleteWebhookConfig();
```


This code is also provided in [RetrieveWebhooks.php](examples/webhooks/RetrieveWebhooks.php),
[SetWebhooks.php](examples/webhooks/SetWebhooks.php),
and [DeleteWebhooks.php](examples/webhooks/DeleteWebhooks.php)


Retrieve Messages From Webhook Queues
----------

Instead of using a public webhook endpoint you can set up your account to put webhooks in a queue system, which can be accessed using this API.
Webhooks will be placed in the queue as soon as the waiver is finished processing and will remain there for 7 days.
Our webhook queues also have support for separate queues for every template if you wish.
To view your current webhook settings, we first need to set a Smartwaiver object.
Make sure to put in your account's API Key where it says `[INSERT API KEY]`

```php
// The API Key for your account
$apiKey = '[INSERT API KEY]';

// Set up your Smartwaiver connection using your API Key
$sw = new Smartwaiver($apiKey);
```

First we can see how many webhook queues we have set up!

```php
// Get the current webhook queue information
$queues = $sw->getWebhookQueues();
```

This returns a nice obejct that has information about how many messages are in our queues.
If our queues don't exist yet, there won't be any information!

```php
// Access the account level webhook information
if (is_null($queues->accountQueue)) {
    echo 'Account Queue: N/A' . PHP_EOL;
} else {
    echo 'Account Queue:' . PHP_EOL;
    echo "\tTotal Messages: " . $queues->accountQueue->messagesTotal . PHP_EOL;
    echo "\tMessages Not Visible: " . $queues->accountQueue->messagesNotVisible . PHP_EOL;
    echo "\tMessages Delayed: " . $queues->accountQueue->messagesDelayed . PHP_EOL;
}

// Access the template level webhook information
foreach ($queues->templateQueues as $templateId => $templateQueue) {
    echo 'Template Queue (' . $templateId . '):' . PHP_EOL;
    echo "\tTotal Messages: " . $queues->accountQueue->messagesTotal . PHP_EOL;
    echo "\tMessages Not Visible: " . $queues->accountQueue->messagesNotVisible . PHP_EOL;
    echo "\tMessages Delayed: " . $queues->accountQueue->messagesDelayed . PHP_EOL;
}
```

Now, if we have some messages in our queue (or not) we can poll to retrieve a message.

```php
// Retrieve a message from the account queue
$message = $sw->getWebhookQueueAccountMessage();
```

Or if we want to get a message from a template queue:

```php
// The Unique ID of the waiver template
$templateId = '[INSERT TEMPLATE ID]';

// Retrieve a message from the template queue
$message = $sw->getWebhookQueueAccountMessage($templateId);
```

This will retrieve one message and we can then access the body of the message.
If there are no messages a null object will be returned.

```php
// Check for empty queue
if (is_null($message)) {
    echo 'No messages in account queue.' . PHP_EOL;
    exit;
}

// Print out message info
echo 'Message in Account Queue' . PHP_EOL;
echo "\tMessage ID: " . $message->messageId . PHP_EOL;
echo "\tMessage Payload: " . PHP_EOL;
echo "\t\tWaiver ID: " . $message->body->uniqueId . PHP_EOL;
echo "\t\tEvent: " . $message->body->event . PHP_EOL;
```

Now that we have processed the message we can delete it.

```php
$delete = $sw->deleteWebhookQueueAccountMessage($message->messageId);

echo 'Deletion Success: ' . ($delete->success ? 'true' : 'false') . PHP_EOL;
```

If we instead wanted to delete it while we retrieved (not recommended since if errors occur in your processing that webhook is lost) we could do this:

```php
// Optionally we can delete the message when we retrieve it, by passing a delete flag
$message = $sw->getWebhookQueueAccountMessage(true);
```

This code is also provided in [RetrieveWebhookQueues.php](examples/queues/RetrieveWebhookQueues.php), RetrieveTemplateMessage.php](examples/queues/RetrieveTemplateMessage.php)
and [RetrieveAccountMessage.php](examples/queues/RetrieveAccountMessage.php)


Exception Handling
==========

Exceptions in this SDK are grouped into three different types.
 * A <b>SmartwaiverSDKException</b> occurs when the SDK itself encounters a problem.
   Examples of this include problems connecting to the API server, an unexpected response from the API server, bad input data, etc.
 * A <b>SmartwaiverHTTPException</b> occurs when the API encounters an error and properly relays that information back.
   Examples of this include '401 Unauthorized' or '404 Not Found' errors.
 * A <b>SmartwaiverRateLimitException</b> occurs when your account hits the rate limit.
   Info is contained in the object to figure out how long to wait before trying again.

Note that a <b>SmartwaiverRateLimitException</b> is a type <b>SmartwaiverHTTPException</b>, which is a type of <b>SmartwaiverSDKException</b> so it is possible to catch all possible exceptions at the same time.

Here is an example of catching an HTTP exception. First we set up the Smartwaiver account:

```php
// The API Key for your account
$apiKey = '[INSERT API KEY]';

// Set up your Smartwaiver connection using your API Key
$sw = new Smartwaiver($apiKey);
```

Next, we attempt to get a waiver that does not exist:

```php
// The Waiver ID to access
$waiverId = 'InvalidWaiverId';

// Try to get the waiver object
$waiver = $sw->getWaiver($waiverId);
```

This will throw an exception because a waiver with that ID does not exist. So let's change the code to catch that exception:

```php
try
{
    // Try to get the waiver object
    $waiver = $sw->getWaiver($waiverId);
}
catch (SmartwaiverHTTPException $se)
{
    // Print out that we encountered an error
    echo 'Error retrieving waiver from API server...' . PHP_EOL . PHP_EOL;
}
```

But there is lot's of useful information in the exception object. Let's print some of that out too:

```php
// The code will be the HTTP Status Code returned
echo 'Error Code: ' . $se->getCode() . PHP_EOL;
// The message will be informative about what was wrong with the request
echo 'Error Message: ' . $se->getMessage() . PHP_EOL . PHP_EOL;

// Also included in the exception is the header information returned about
// the response.
$responseInfo = $se->getResponseInfo();
echo 'API Version: ' . $responseInfo['version'] . PHP_EOL;
echo 'UUID: ' . $responseInfo['id'] . PHP_EOL;
echo 'Timestamp: ' . $responseInfo['ts'] . PHP_EOL;
```

The code provided here is also combined in to one example in [ExceptionHandling.php](examples/intro/ExceptionHandling.php)

The <b>SmartwaiverRateLimitException</b> is a special type of <b>SmartwaiverHTTPException</b> and provides some additional information.
The exception can be caught in exactly the same way:

```php
try
{
    // Try to get the waiver object
    $waiver = $sw->getWaiver($waiverId);
}
catch (SmartwaiverRateLimitException $se)
{
    // The number of requests that have taken place in the current window
    echo 'Number of requests: ' . $se->requests . PHP_EOL;
    // The max number of requests for routes of this type in the current window
    echo 'Max requests: ' . $se->max . PHP_EOL;
    // The number of seconds to wait until trying again
    echo 'Retry after: ' . $se->retryAfter . PHP_EOL;
}
```

And the exception object has some additional information, seen being printed out above.

Status Codes
----------

The code of the exception will match the HTTP Status Code of the response and the message will be an informative string informing on what exactly was wrong with the request.

Possible status codes and their meanings:

| Status Code | Error Name            | Description                                                                                                                       |
| ----------- | --------------------- | --------------------------------------------------------------------------------------------------------------------------------- |
| 400         | Parameter Error       | Indicates that something was wrong with the parameters of the request (e.g. extra parameters, missing required parameters, etc.). |
| 401         | Unauthorized          | Indicates the request was missing an API Key or contained an invalid API Key.                                                     |
| 402         | Data Error            | Indicates that the parameters of the request was valid, but the data in those parameters was not.                                 |
| 404         | Not Found             | Indicates that whatever was being searched for (specific waiver, etc.) could not be found.                                        |
| 406         | Wrong Content Type    | Indicates that the Content Type of the request is inappropriate for the request.                                                  |
| 429         | Too Many Requests     | Indicates that your account rate limit has been hit and you should cease making requests.                                         |
| 500         | Internal Server Error | Indicates that the server encountered an internal error while processing the request.                                             |

Advanced
==========

This section contains notes about several more ways to use the SDK that are slightly more low level.

Raw Responses
----------
If you do not wish to use the Smartwaiver object types to facilitate easy use of the data you can also access the raw response from the API server.

Here is an example of getting the raw response from the server for retrieving a list of waiver summaries:

```php
// The API Key for your account
$apiKey = '[INSERT API KEY]';

// Set up your Smartwaiver connection using your API Key
$sw = new Smartwaiver($apiKey);

// Get a list of all signed waivers for this account
$response = $sw->getWaiverSummariesRaw();

// The response object has two variables, status code and response body
echo 'Status Code: ' . $response->statusCode . PHP_EOL;
echo 'Body: ' . $response->body . PHP_EOL;
```

All the standard methods have a 'Raw' counterpart that just has 'Raw' added to the function name.

The code provided here is also in [RawResponses.php](examples/advanced/RawResponses.php)

URL Generation
----------

If you would like handle all aspects of the request's yourself, you can simply use <b>SmartwaiverRoutes</b> class to generate the approriate URLs for your requests.

For example, to create the URL to list all templates is only one line:

```php
SmartwaiverRoutes::getWaiverTemplates();
```

For the list of possible routes see [SmartwaiverRoutes.php](src/Smartwaiver/SmartwaiverRoutes.php)


CHANGELOG.md
# Release Notes

## v4.3.1 (2019-01-15)

- Add standard fields to dynamic document data object
- Add route to delete account webhook setting

## v4.3.0 (2018-12-19)

- Add ability to create and process dynamic documents

## v4.2.9 (2018-09-27)

- Added vanity URLs to template objects

## v4.2.8 (2018-09-05)

- Added webhook queues

## v4.2.4 (2018-05-24)

- Added ability to access custom fields by GUID
- Added signatures route

## v4.2.3 (2018-01-08)

### Added
- Added exception for rate limiting
- Added flags to waiver results ([Example](https://github.com/smartwaivercom/php-sdk#retrieve-a-specific-waiver))
- Added searching of primary tag to search and waivers routes ([Example](https://github.com/smartwaivercom/php-sdk#list-all-signed-waivers))
- Added field indicating number of photos to waiver body ([Example](https://github.com/smartwaivercom/php-sdk#retrieve-a-specific-waiver))
- Added new route to retrieve waiver photos ([Example](https://github.com/smartwaivercom/php-sdk#retrieve-photos-on-a-waiver))
- Added new search route ([Example](https://github.com/smartwaivercom/php-sdk#search-for-waivers))
- Added first and last name query parameters to waivers route ([Example](https://github.com/smartwaivercom/php-sdk#list-all-signed-waivers))


## v4.2.4 (2023-11-15)

- Updated PHP to use 7.4
- Updated Guzzle to use ^7.0


examples/intro/ExceptionHandling.php
<?php
/**
 * Copyright 2018 Smartwaiver
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

require_once __DIR__ . '/../../../../autoload.php';

use Smartwaiver\Smartwaiver;
use Smartwaiver\Exceptions\SmartwaiverHTTPException;

// The API Key for your account
$apiKey = '[INSERT API KEY]';

// The Waiver ID to access
$waiverId = 'InvalidWaiverId';

// Set up your Smartwaiver connection using your API Key
$sw = new Smartwaiver($apiKey);

try
{
    // Try to get the waiver object
    $waiver = $sw->getWaiver($waiverId);
}
catch (SmartwaiverHTTPException $se)
{
    // SmartwaiverHTTPException will be thrown for any errors returned by the
    // API in a RESTful way.
    // Examples include: 404 Not Found, 401 Not Authorized, etc.
    echo 'Error retrieving waiver from API server...' . PHP_EOL . PHP_EOL;

    // The code will be the HTTP Status Code returned
    echo 'Error Code: ' . $se->getCode() . PHP_EOL;
    // The message will be informative about what was wrong with the request
    echo 'Error Message: ' . $se->getMessage() . PHP_EOL . PHP_EOL;

    // Also included in the exception is the header information returned about
    // the response.
    $responseInfo = $se->getResponseInfo();
    echo 'API Version: ' . $responseInfo['version'] . PHP_EOL;
    echo 'UUID: ' . $responseInfo['id'] . PHP_EOL;
    echo 'Timestamp: ' . $responseInfo['ts'] . PHP_EOL;
}
